src/app/policies/views/PolicyListView.vue

```diff
@@ -15,95 +15,222 @@
         policy: '',
       }"
     >
-      <AppView>
-        <template #title>
-          <h2>
-            <RouteTitle
-              :title="t('policies.routes.items.title')"
-            />
-          </h2>
-        </template>
-
-        <!--
-          Load in all the potential policy types.
-          This means we can find an valid 'selected' type.
-          It’s also used for the list of policy types.
-        -->
-        <DataSource
-          v-slot="{ data: policyTypesData, error: policyTypesError }: PolicyTypeCollectionSource"
-          :src="`/*/policy-types`"
+      <DataCollection
+        v-slot="{ items: types }"
+        :predicate="(policyType) => typeof policyType !== 'undefined' && policyType.path === route.params.policyPath"
+        :find="true"
+        :items="props.policyTypes ?? [undefined]"
+      >
+        <template
+          v-for="policyType in [types[0]!]"
+          :key="policyType"
```

Comment:
**Suggestion**: This is a lot of unnecessary abstraction here. First of all, policy types can’t be empty. They can error, but they can’t be empty. Secondly, why are we using a Vue component for the task of finding an item in a list? A simple computed property does this job so much better. Besides which, using a component intended to show a collection of things with a empty block baked in for something that will 100% of times show one thing and not a collection is really weird as well.

Reply:
> First of all, policy types can’t be empty.

They can be empty before they are loaded:

https://github.com/kumahq/kuma-gui/pull/2250/files#diff-d2633793eb7b7dee98cca8d616577a208c7eba06b7b8f983e4b7c0fce8986a52R104

> Secondly, why are we using a Vue component for the task of finding an item in a list?

Partly because I wanted an alternative to `<template v-for>` which you don't like, and partly because we get an automatic empty state with it.

> Besides which, using a component intended to show a collection of things with a empty block baked in for something that will 100% of times show one thing and not a collection is really weird as well.

See above, it can be empty before they are loaded.

> A simple computed property does this job so much better.

Computed properties don't have automatic empty states

Reply:
> They can be empty before they are loaded:

They’re not empty then; they’re not loaded yet, and more specifically _are being loaded_. From a user perspective, you never expect to see an empty state for policy types.

> Partly because I wanted an alternative to `<template v-for>` which you don't like, and partly because we get an automatic empty state with it.

It looks like this neither needs an empty state nor is it able to avoid using the template v-for hack as one is following it immediately.

There are two pieces here why I think DataCollection is inadequate for what you’re trying to get to here:

- Passing in `[undefined]` when `props.policyTypes` is not set. Of course DataCollection expects an array of things. Providing it with `[undefined]` says both "I can’t provide it with a list of the right things right now" and "I'm still gonna provide it with a list of wrong things". This can only happen while still loading policy types, but there is a DataLoader that already covers the loading/empty scenario further up. That already enables you to expect a non-empty and set `props.policyTypes`. That’s handled in the current component in `master` so I don’t see how we can’t keep this convenience. You also have to guard against `undefined` in the `predicate` prop which is another signal that something is wrong here. The response from `/policies` can’t have `undefined` policy type definitions.
- There’s now a new instance of the `v-for` variable setting template hack with a type assertion on accessing `types[0]!`.

> Computed properties don't have automatic empty states

Of course not. The point? You can add an empty state if you need one (which you don’t here). That has nothing to do with computed properties.

---

So, from looking at this, neither does this part of the code need an empty state as that is already handled further up when loading the data nor does it need any of the patch work of selecting the current policy type. This component should be able to expect (1) a loaded set of `PolicyType`s and (2) the current `PolicyType` object as found from the current policy type path from the path segment. That’s an ideal use case for a computed property.

Reply:
Why do we want to show the empty state for an unknown policy type? The current behavior (showing an error) state is what I’d expect here.

Reply:
It’s expected that it does that. You manually typed in a URL to a policy type that we don’t know about (and that you can’t get to from anywhere in the UI) and you get a generic 404 error. That’s about to be expected.

Should we need to handle arbitrary policy types at some point due to some form of plug-and-play policy concept being introduced, many things will have to be rethought and this doesn’t help.

Seeing that the outfall of this is a bunch of convoluted logic in this component that we otherwise wouldn’t need (and actually probably even with it, we could handle this much better), I’m not okay with doing it that way.

To sum up what I expect to happen here:

- for unknown policy types, the view should show an error page as it does in `master`
- the PolicyListView component shouldn’t try and handle empty states and the like for policy type sourcing as that’s already done in PolicyTypeListView.

The rest of the PR looks more or less good to me even though I still kind of question why we need two different views for this. The separation of data-loading view and data-presenting component makes more sense to me.

Reply:
> It’s expected that it does that. 

🤔  I'd say its a bug. It gets even weirder when the CircuitBreaker item isn't in the left hand menu at all, and then it shows me an "About CircuitBreaker" intro text. Why do we show a piece of intro text for something that doesn't even exist? It's not that there aren't any, it's that that don't exist. As a user I'd be like "what is a circuit breaker and why are you showing me this when I asked for something entirely different?

> Should we need to handle arbitrary policy types at some point due to some form of plug-and-play policy concept being introduced, many things will have to be rethought and this doesn’t help.

As far as I am aware we have been handling these cases for the very very long time in the code. How come we shouldn't handle them here? We may aswell just pretend the list of policies is hardcoded in the code? (obvs I don't thin kwe should do that and hence this fix)

Reply:
btw is this a suggestion or a blocker?

Reply:
So just to note what was said in the sync.

We are going to go with what is implemented in this PR, but we are going to change the "No Data" text to say something like "This policy doesn't exist"..

Reply:
I went with this:

![Screenshot 2024-03-12 at 11 25 53](https://github.com/kumahq/kuma-gui/assets/554604/eb71f5fb-12c7-4c49-a4ed-44bbcbb5e831)

The text is also in i18n, so super easy to change if we want to. Will push this on the end now and resolve this conversation.

Reply:
I added one more commit here to move the custom text to the body rather than the title, that way it looks much more like everywhere else where we do this:

### Non existing policy type

![Screenshot 2024-03-13 at 10 47 54](https://github.com/kumahq/kuma-gui/assets/554604/7f705b3a-3cff-45b9-bfd6-6026add39915)

### Existing empty state for non existing policies

![Screenshot 2024-03-13 at 10 47 45](https://github.com/kumahq/kuma-gui/assets/554604/0565bfb0-a4c2-4199-80d7-2a188275f0c7)